### PR TITLE
Add StateMachineTest framework for JUnit 5

### PR DESCRIPTION
**Pull request by divang**
**Summary**

Introduces a lightweight State Machine Testing (MBT) framework that randomly explores valid JDBC state transitions to discover edge-case bugs. Designed as a simpler, debuggable, seed-reproducible alternative to the existing MBT framework.

**Motivation**

**Hand-written tests miss paths**
We only test sequences we think of. Real applications produce unexpected combinations like insert, rollback, insert, commit, rollback, insert, commit -- nobody writes a test for that, yet it may crash the driver.
**The existing MBT framework is hard to debug**
It uses reflection, annotations, and an internal scheduler. When a test fails, stepping through in a debugger is impractical because execution jumps across framework internals.
**Failures are not reproducible**
The existing framework has no seed-based replay. A nightly test fails, but running it again produces a different random sequence -- the bug disappears. Engineers waste time trying to reproduce instead of fixing.

**Key Design Decisions**

**Seed-based reproducibility** -- Same seed produces the same action sequence on any machine. Failed CI? Paste the seed, set a breakpoint, reproduce on the first try.
**Weighted random selection** -- executeUpdate (weight 15) fires 3x more often than setAutoCommit (weight 5), simulating realistic workloads.
**No third-party dependencies** -- Pure JUnit 5 and standard library. No external libraries, no annotations, no reflection.

**What This PR Adds**

**Core Framework (statemachinetest/core/)** -- 5 classes, around 200 lines
**Action** -- Base class with canRun() (precondition), run() (execute), and weight (selection probability)
**Engine** -- Execution loop with weighted random selection, seed, timeout
**StateMachineTest** -- Shared state map, action registry, seeded Random
**StateKey** -- Interface for type-safe enum state keys (no raw strings)
**Result** -- Outcome containing success, actionCount, seed, log, error

**Transaction Tests (statemachinetest/transaction/)**
TransactionState enum with keys CONN, AUTO_COMMIT, CLOSED, TABLE_NAME
6 actions: setAutoCommit(true/false), commit, rollback, executeUpdate, executeQuery
TransactionStateTest -- 50 random actions against real SQL Server

**ResultSet Tests (statemachinetest/resultset/)**
ResultSetState enum with keys RS, CLOSED, ON_VALID_ROW
6 actions: next, previous, first, last, absolute, getString
ResultSetStateTest -- 50 random actions against real SQL Server

**Testing**

2 test classes, 50 actions each, run against real SQL Server

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/core/StateMachineTest.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+                            selected = a;
+                            break;
+                        }
+                    }
+
+                    count++;
+                    selected.run();
+                    log.add(selected.name);
+                    System.out.printf("[%3d] %-20s | %s%n", count, selected.name, sm.state);
+                }
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/core/StateMachineTest.java]: Currently we are evaluating this test framework, so let the verbose logging in place.

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 2]
```diff
@@ -0,0 +1,2212 @@
...
+     */
+    private static void createTestTable(Connection conn) throws SQLException {
+        try (Statement stmt = conn.createStatement()) {
+            TestUtils.dropTableIfExists(TABLE_NAME, stmt);
+            stmt.execute("CREATE TABLE " + TABLE_NAME + " (id INT PRIMARY KEY, value INT)");
+            stmt.execute("INSERT INTO " + TABLE_NAME + " VALUES (1, 100)");
+        }
+    }
+
+    // ==================== SIMULATED TRANSACTION TESTS (No DB needed) ====================
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: Cleaned up the simulated test cases.

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 3]
```diff
@@ -0,0 +1,2212 @@
...
+ * See the LICENSE file in the project root for more information.
+ */
+package com.microsoft.sqlserver.jdbc.statemachinetest.transaction;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Savepoint;
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: Cleaned up

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 4]
```diff
@@ -0,0 +1,2212 @@
...
+            return !sm.is("closed");
+        }
+
+        @Override
+        public void run() throws SQLException {
+            Connection conn = (Connection) sm.get("conn");
+            try (Statement stmt = conn.createStatement()) {
+                int newValue = (int) (Math.random() * 1000);
+                int rows = stmt.executeUpdate("UPDATE " + TABLE_NAME + " SET value = " + newValue + " WHERE id = 1");
+                System.out.println("  executeUpdate -> " + rows + " rows, value=" + newValue);
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: Cleaned up.

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/core/StateMachineTest.java [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+     */
+    public List<Action> getValidActions() {
+        List<Action> valid = new ArrayList<>();
+        for (Action a : actions) {
+            try {
+                if (a.canRun()) {
+                    valid.add(a);
+                }
+            } catch (Exception e) {
+                // Skip actions that throw during precondition check
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/core/StateMachineTest.java]: Handled

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 6]
```diff
@@ -0,0 +1,1606 @@
...
+        @Override
+        public void run() {
+            sm.set("row", sm.getInt("rowCount"));
+        }
+    }
+    
+    static class AbsoluteAction extends Action {
+        private StateMachineTest sm;
+        private Random random = new Random();
+
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: Cleaned up

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 7]
```diff
@@ -0,0 +1,1606 @@
...
+                newRow = Math.max(0, rows + 1 + target);
+            }
+            sm.set("row", newRow);
+        }
+    }
+    
+    static class RelativeAction extends Action {
+        private StateMachineTest sm;
+        private Random random = new Random();
+
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: Cleaned up

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 8]
```diff
@@ -0,0 +1,1606 @@
...
+        
+        @Override
+        public void run() {
+            ResultSet rs = (ResultSet) sm.get("rs");
+            try {
+                String name = rs.getString("name");
+                System.out.println("  getString('name') -> " + name);
+            } catch (SQLException e) {
+                System.out.println("  getString failed: " + e.getMessage());
+            }
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: Handled

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 9]
```diff
@@ -0,0 +1,1606 @@
...
+        try (Statement stmt = connection.createStatement()) {
+            TestUtils.dropTableIfExists(TABLE_NAME, stmt);
+            stmt.execute("CREATE TABLE " + TABLE_NAME + " (id INT PRIMARY KEY, name VARCHAR(50), value INT)");
+            for (int i = 1; i <= 10; i++) {
+                stmt.execute("INSERT INTO " + TABLE_NAME + " VALUES (" + i + ", 'Row" + i + "', " + (i * 10) + ")");
+            }
+        }
+    }
+    
+    // ==================== SIMULATED RESULTSET (No DB needed) ====================
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: cleaned up

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 10]
```diff
@@ -0,0 +1,2212 @@
...
+        }
+    }
+
+    // FX _modelcreateStatement(type, concurrency): weight=10
+    static class CreateStatementWithTypeAction extends Action {
+        private StateMachineTest sm;
+        private int[] cursorTypes = {ResultSet.TYPE_FORWARD_ONLY, ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.TYPE_SCROLL_INSENSITIVE};
+        private int[] concurrencyModes = {ResultSet.CONCUR_READ_ONLY, ResultSet.CONCUR_UPDATABLE};
+        private java.util.Random random = new java.util.Random();
+
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: cleaned up

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 11]
```diff
@@ -0,0 +1,2212 @@
...
+            return !sm.is("closed") && !sm.is("autoCommit") && savepoints != null && !savepoints.isEmpty();
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void run() {
+            List<String> savepoints = (List<String>) sm.get("savepoints");
+            // Rollback to last savepoint
+            String sp = savepoints.get(savepoints.size() - 1);
+            // Remove all savepoints after this one (they're invalidated)
```
[the remark this answers is not in the document]
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: cleaned up

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+ * the simple StateMachineTest framework.
+ * 
+ * Test scenarios covered:
+ * - Scrollable cursor navigation (next, previous, first, last, absolute)
+ * - Data retrieval (getString)
+ */
+@Tag(Constants.stateMachine)
+public class ResultSetStateTest extends AbstractTest {
+
+    private static final String TABLE_NAME = AbstractSQLGenerator.escapeIdentifier("SM_ResultSet_Test");
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: Use RandomUtil.getIdentifier() to get a unique table name for each test run.
```suggestion
    private static final String TABLE_NAME = AbstractSQLGenerator.escapeIdentifier(RandomUtil.getIdentifier("SM_ResultSet_Test"));
```

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+ * 
+ * Test scenarios covered:
+ * - setAutoCommit(true/false) - Toggle auto-commit mode
+ * - commit() - Commit transaction (requires autoCommit=false)
+ * - rollback() - Rollback transaction (requires autoCommit=false)
+ */
+@Tag(Constants.stateMachine)
+public class TransactionStateTest extends AbstractTest {
+
+    private static final String TABLE_NAME_CONST = AbstractSQLGenerator.escapeIdentifier("SM_Transaction_Test");
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: Use RandomUtil.getIdentifier() to get a unique table name for each test run.
```suggestion
    private static final String TABLE_NAME_CONST = AbstractSQLGenerator.escapeIdentifier(RandomUtil.getIdentifier("SM_Transaction_Test"));
```

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+            sm.setState(ON_VALID_ROW, false);
+
+            sm.addAction(new NextAction(sm));
+            sm.addAction(new PreviousAction(sm));
+            sm.addAction(new FirstAction(sm));
+            sm.addAction(new LastAction(sm));
+            sm.addAction(new AbsoluteAction(sm));
+            sm.addAction(new GetStringAction(sm));
+
+            Result result = Engine.run(sm).withMaxActions(50).withSeed(99999).execute();
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: I'm confused. I thought the idea was to randomize the actions across runs. A static seed will produce the same path every time, no? Don't we want to randomize the seed (and log it) so that we vary coverage while logging for reproducibility/debugging?
```suggestion
            Result result = Engine.run(sm).withMaxActions(50).execute();
```

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+            sm.setState(TABLE_NAME, TABLE_NAME_CONST);
+
+            sm.addAction(new SetAutoCommitFalseAction(sm));
+            sm.addAction(new SetAutoCommitTrueAction(sm));
+            sm.addAction(new CommitAction(sm));
+            sm.addAction(new RollbackAction(sm));
+            sm.addAction(new ExecuteUpdateAction(sm));
+            sm.addAction(new SelectAction(sm));
+
+            Result result = Engine.run(sm).withMaxActions(50).withSeed(54321).execute();
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/transaction/TransactionStateTest.java]: Same question here with the static seed...
```suggestion
            Result result = Engine.run(sm).withMaxActions(50).execute();
```

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java (hunk 14 above)
Comment (divang) [reply src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java]: Yes, that was the sample one. It should be randomized.

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        try (Statement stmt = connection.createStatement()) {
+            TestUtils.dropTableIfExists(TABLE_NAME, stmt);
+            stmt.execute("CREATE TABLE " + TABLE_NAME + " (id INT PRIMARY KEY, name VARCHAR(50), value INT)");
+            for (int i = 1; i <= 10; i++) {
+                stmt.execute("INSERT INTO " + TABLE_NAME + " VALUES (" + i + ", 'Row" + i + "', " + (i * 10) + ")");
+            }
+        }
+    }
+
+    @Test
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java:77]: Add legacyFX test tag to all new tests

File: src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    @DisplayName("FX Model: Real Database - Scrollable Sensitive Cursor")
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/jdbc/statemachinetest/resultset/ResultSetStateTest.java:78]: legacyFX test tag

File: src/test/java/com/microsoft/sqlserver/testframework/Constants.java [hunk 18]
```diff
@@ -57,6 +58,7 @@ private Constants() {}
     public static final String JSONTest = "JSONTest";
     public static final String CodeCov = "CodeCov";
     public static final String PrepareMethodUseTempTableScopeTest = "PrepareMethodUseTempTableScopeTest";
+    public static final String stateMachine = "stateMachine";
```
Comment (machavan) [inline src/test/java/com/microsoft/sqlserver/testframework/Constants.java]: what is this defined for ?

Comment (machavan) [reply src/test/java/com/microsoft/sqlserver/testframework/Constants.java]: lets use legacyFX

Comment (divang) [reply src/test/java/com/microsoft/sqlserver/testframework/Constants.java]: not required, it should be legacyFX

**Review by Ananya2 — DISMISSED** [review]
